File: src/ai_agent_task/app/routes/tasks.py
from fastapi import APIRouter
from ..schemas import TaskCreate, TaskUpdate

tasks = []

router = APIRouter(prefix="/tasks", tags=["tasks"])


@router.get("/tasks")
def get_tasks():
    return tasks


@router.post("/tasks")
def create_task(task: TaskCreate):
    new_task = {
        "id": len(tasks) + 1,
        "title": task.title,
        "date": task.date,
        "completed": task.completed,
    }

    tasks.append(new_task)

    return new_task

@router.patch("/tasks/{task_id}")
def update_task(task_id: int, update: TaskUpdate):
    for task in tasks:
        if task["id"] == task_id:
            if update.title is not None:
                task["title"] = update.title

            if update.completed is not None:
                task["completed"] = update.completed

            if update.date is not None:
                task["date"] = update.date

            return task

    return {"error": "Task not found"}


@router.delete("/tasks/{task_id}")
def delete_task(task_id: int):
    for task in tasks:
        if task["id"] == task_id:
            tasks.remove(task)
            return {"message": "Task deleted"}

    return {"error": "Task not found"}

File: frontend/src/api/tasks.js
const API_URL = 'http://127.0.0.1:8000'

export async function getTasks() {
  const response = await fetch(`${API_URL}/tasks`)

  if (!response.ok) {
    throw new Error('Failed to load tasks')
  }

  return response.json()
}

export async function createTask(title, date) {
  const response = await fetch(`${API_URL}/tasks`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      title: title,
      date: date || null,
      completed: false,
    }),
  })
  
  if (response.status === 422) {
    throw new Error('Too large input')
  }

  if (!response.ok) {
    throw new Error('Failed to create task')
  }
}

export async function updateTask(taskId, updates) {
  const response = await fetch(`${API_URL}/tasks/${taskId}`, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(updates),
  })

  if (!response.ok) {
    throw new Error('Failed to update task')
  }
}

export async function deleteTask(taskId) {
  const response = await fetch(`${API_URL}/tasks/${taskId}`, {
    method: 'DELETE',
  })

  if (!response.ok) {
    throw new Error('Failed to delete task')
  }
}